File: examples/canonical-gcp-functions/http-stripe-webhook/typescript/src/handler.ts
import crypto from "crypto";

export type Clients = {
  storeRawEvent: (eventId: string, payload: unknown) => Promise<boolean>;
  enqueueFulfillment: (eventId: string, payload: unknown) => Promise<void>;
  getSigningSecret: () => Promise<string>;
};

export async function handler(req: any, clients?: Clients) {
  const started = Date.now();
  const activeClients = clients ?? gcpClients();
  const body = typeof req.body === "string" ? req.body : JSON.stringify(req.body ?? {});
  const payload = JSON.parse(body);
  const eventId = payload.id;
  const signature = req.headers?.["stripe-signature"] ?? req.headers?.["Stripe-Signature"];
  const signingSecret = await activeClients.getSigningSecret();
  if (!verifyStripeSignature(body, signature, signingSecret)) {
    log(started, eventId, eventId, "DROP", "invalid stripe signature");
    return { statusCode: 400, body: "invalid signature" };
  }
  const inserted = await activeClients.storeRawEvent(eventId, payload);
  if (inserted) {
    await activeClients.enqueueFulfillment(eventId, payload);
  }
  log(started, eventId, eventId, inserted ? "ALLOW" : "DROP", inserted ? "accepted" : "duplicate");
  return { statusCode: 202, body: JSON.stringify({ accepted: true, duplicate: !inserted }) };
}

export function verifyStripeSignature(body: string, signature: string | undefined, signingSecret: string): boolean {
  if (!signature) return false;
  const expected = crypto.createHmac("sha256", signingSecret).update(body).digest("hex");
  return signature.includes(expected);
}

function gcpClients(): Clients {
  throw new Error("GCP clients are bound by deployment; tests inject clients.");
}

function log(started: number, correlationId: string, dedupeKey: string, decision: string, msg: string) {
  console.log(JSON.stringify({
    timestamp: new Date().toISOString(),
    severity: "INFO",
    msg,
    trace_id: process.env.TRACEPARENT ?? "local-trace",
    request_id: correlationId,
    correlation_id: correlationId,
    provider: "gcp",
    runtime_tier: "function",
    function_name: process.env.FUNCTION_TARGET ?? "accb-dev-gcp-stripe-handler",
    dedupe_key: dedupeKey,
    decision,
    latency_ms: Date.now() - started
  }));
}

File: examples/canonical-gcp-functions/firebase-onboarding/python/src/handler.py
from __future__ import annotations

import json, os, time
from dataclasses import dataclass
from typing import Any, Protocol

class FirestoreClient(Protocol):
    def claim_user_event(self, dedupe_key: str, payload: dict[str, Any]) -> bool: ...
    def write_profile(self, uid: str, payload: dict[str, Any]) -> None: ...
class PubSubClient(Protocol):
    def publish(self, topic: str, payload: dict[str, Any]) -> None: ...
@dataclass
class Clients:
    firestore: FirestoreClient
    pubsub: PubSubClient

def handle_firebase_user(event: dict[str, Any], clients: Clients | None = None) -> dict[str, Any]:
    started = time.time()
    data = event.get("data", {})
    uid = data["uid"]
    email = data.get("email", "")
    event_id = event.get("id", uid)
    dedupe_key = f"{uid}:{event_id}"
    clients = clients or _gcp_clients()
    if not clients.firestore.claim_user_event(dedupe_key, data):
        _log(started, event_id, dedupe_key, "DROP", "duplicate firebase onboarding")
        return {"ok": True, "duplicate": True, "decision": "DROP"}
    profile = {"uid": uid, "email": email, "status": "ONBOARDING_STARTED"}
    clients.firestore.write_profile(uid, profile)
    clients.pubsub.publish(os.getenv("ONBOARDING_TOPIC", "accb-dev-gcp-firebase-onboarding"), profile)
    _log(started, event_id, dedupe_key, "ALLOW", "firebase onboarding started")
    return {"ok": True, "duplicate": False, "decision": "ALLOW", "profile": profile}

def _gcp_clients() -> Clients:
    raise RuntimeError("GCP clients are bound by deployment; tests inject clients.")

def _log(started: float, event_id: str, dedupe_key: str, decision: str, msg: str) -> None:
    print(json.dumps({"timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()), "severity": "INFO", "msg": msg, "trace_id": os.getenv("TRACEPARENT", "local-trace"), "request_id": event_id, "correlation_id": event_id, "provider": "gcp", "runtime_tier": "function", "function_name": os.getenv("FUNCTION_TARGET", "accb-dev-gcp-firebase-onboarding"), "dedupe_key": dedupe_key, "decision": decision, "latency_ms": int((time.time() - started) * 1000)}))
